### PR TITLE
Shell script and chunk analysis

### PR DESCRIPTION
**Pull request by bennyrubanov**

### CODE REVIEW

**Comment by EllAchE** [comment]
> _ No description provided. _

you have 2 shell PRs

**Comment by bennyrubanov** [comment]
next steps:
    - figure out why getting errors while reading through the files being created when streaming_partial_decompresser is run. most likely because parts of the games are getting cut off and disrupting fileReader. need to implement catches for that
        - can potentially only write to new file when a game is not cut off, if can pass data back to the decompression script?
    - see if i can implement an “analysis heap” or “stack wait list” so that analyses for each file are waiting to be finished before the file is deleted (for some reason, analysis is still not done before file gets deleted)
    - figure out how to store the output from each analysis to be used as inputs into the final analysis function

**Comment by bennyrubanov** [comment]
ignore old shell PR for now, need this one

**Comment by bennyrubanov** [comment]
@EllAchE need your help mostly with figuring out how to get the "next steps" items done in my comment above. The script for partial decompression + analysis is turning out to be a challenge

File: src/streaming_partial_decompresser.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+                let lastChunkLength = 0;
+                let fileLength = 0;
+
+                result.on('error', (err) => {
+                    return reject(err);
+                });
+
+                result.on('data', async (data) => {
+                    decompressedStream.write(data);
```
Comment (EllAchE) [inline src/streaming_partial_decompresser.js:84]: do you clear the stream? Are you overwriting it? Won't this end up being OOM?

**Review by EllAchE — COMMENTED** [review]
I came here to take a quick look because I figured this opinino would be relevant - I have little context so coming in and trying to figure out what in your code works/deosn't is possibly just as much or more effort as a rewrite.

What would help here is if you can isolate atomic tasks that your code needs to complete and ensure those work. Write unit tests for the working things AND the broken things, then let me take a look at the broken things

**Comment by bennyrubanov** [comment]
> I came here to take a quick look because I figured this opinino would be relevant - I have little context so coming in and trying to figure out what in your code works/deosn't is possibly just as much or more effort as a rewrite.
> 
> What would help here is if you can isolate atomic tasks that your code needs to complete and ensure those work. Write unit tests for the working things AND the broken things, then let me take a look at the broken things

yeah agreed. the issues i'm having are quite complex and i'll take another stab at it myself. if i still can't ill work on isolating tasks and separating out what's broken for you to look at

File: src/streaming_partial_decompresser.js (hunk 1 above)
Comment (bennyrubanov) [reply src/streaming_partial_decompresser.js:84]: decompressedStream.write(data) appends the data to the file, not storing it in memory. shouldn't cause an out-of-memory (OOM) error

the strat is to not stop the stream, write it to a file until the file hits the size limit, then immediately redirect the stream to a new file and start analysis on the old one, rinse and repeat

Comment (bennyrubanov) [reply src/streaming_partial_decompresser.js:84]: so data is continuously being written to disk, not stored in memory. doesn't clear stream 

also trying to implement old file deletion once analysis is done

basically parallel processes: stream write to new file, analysis + deletion of old file

**Comment by EllAchE** [comment]
some merge conflicts here

File: cjsmin/src/chess.ts [hunk 2]
```diff
@@ -1737,7 +1737,7 @@ export class Chess {
         };
         yield {
           move: prettyMove,
-          board: this._board,
+          board: [...this._board],
```
Comment (EllAchE) [inline cjsmin/src/chess.ts:1740]: If we care about performance we should find another way to do this - for now we can leave a TODO

File: cjsmin/src/chess.ts [hunk 3]
```diff
@@ -1787,12 +1787,12 @@ export class Chess {
...
         result.push({
           move: prettyMove,
-          board: this._board,
+          board: [...this._board], // need to copy the board, not pass a reference to it since we are aggregating and doing one game at a time. Copying this is SLOW
         });
       }
     }
 
-    return result;
+    return result.length ? result : [];
```
Comment (EllAchE) [inline cjsmin/src/chess.ts:1795]: hmm I guess we get some invalid strings, should this also be an error case?

File: src/fileReader.ts [hunk 4]
```diff
@@ -13,23 +13,43 @@ export async function* gameChunks(
...
+        }
+      }
+    } else if (line.trim() === '') {
+      // Empty line, do nothing
     } else {
-      // AFAIK this should never happen
-      console.log(`Unknown line: ${line}`);
-      throw new Error(`Unknown line: ${line}`);
+      // Unknown line, ignore the current game
+      console.log(`Unknown line: ${line}, game will be ignored.`);
```
Comment (EllAchE) [inline src/fileReader.ts:49]:
```suggestion
      console.warn(`Unknown line: ${line}, game will be ignored.`);
```

File: src/fileReader.ts [hunk 5]
```diff
@@ -13,23 +13,43 @@ export async function* gameChunks(
...
+  let ignoreGame: boolean = false;
   for await (const line of reader) {
-    // metadata & move lines chunked as a single game
-    if (line.startsWith('[')) {
-      metadata.push(line);
-    } else if (!line) {
-      // empty line, do nothing
+    if (line.startsWith('[Event')) {
+      // Start of a new game
+      ignoreGame = false;
```
Comment (EllAchE) [inline src/fileReader.ts:21]: if we are ignoring game couldn't we just return? Then the generator will yield nothing. And it'll simplify this logic

File: src/metrics/captures.ts [hunk 6]
```diff
@@ -148,6 +208,22 @@ export class KDRatioMetric implements Metric {
...
+          this.KDAssistsValuesMap[move.uas].valueKills++;
+        } else if (move.capture.type === 'n' || move.capture.type === 'b') {
+          this.KDAssistsValuesMap[move.uas].valueKills += 3;
+        } else if (move.capture.type === 'k') {
+          this.KDAssistsValuesMap[move.uas].valueKills += 4; // king's fighting value valued around 4 points in https://en.wikipedia.org/wiki/Chess_piece_relative_value
+        } else if (move.capture.type === 'r') {
+          this.KDAssistsValuesMap[move.uas].valueKills += 5;
+        } else if (move.capture.type === 'q') {
+          this.KDAssistsValuesMap[move.uas].valueKills += 9;
+        }
```
Comment (EllAchE) [inline src/metrics/captures.ts:224]: minor but a map of type --> value might be faster than the worst-case 5 boolean checks, and we could add default behavior. A switch statement would apply

File: src/metrics/distances.ts [hunk 7]
```diff
@@ -11,6 +11,7 @@ export class MoveDistanceMetric implements Metric {
   minAvgDistance: number;
   totalDistance: number;
   maxSingleGameTotal: number;
+  singleGameMaxPiece: string[];
```
Comment (EllAchE) [inline src/metrics/distances.ts:14]: why not use UAP etc. here

File: src/metrics/distances.ts [hunk 8]
```diff
@@ -135,6 +134,10 @@ export class MoveDistanceMetric implements Metric {
     // add the single game aggregates to the state object
     for (const uas of Object.keys(singleGameMap)) {
       this.distanceMap[uas].total += singleGameMap[uas].total;
+      const gameLink = metadata.find((item) => item.startsWith('[Site "'))
+      ?.replace('[Site "', '')
+      ?.replace('"]', '');
```
Comment (EllAchE) [inline src/metrics/distances.ts:139]: metadata could be an object/map if we want?

File: src/metrics/distances.ts [hunk 9]
```diff
@@ -145,9 +148,10 @@ export class MoveDistanceMetric implements Metric {
       if (singleGameMap[uas].total > this.pieceMaxes.distance) {
         this.pieceMaxes.uasArray = [uas as UASymbol];
         this.pieceMaxes.distance = singleGameMap[uas].total;
-        // this.singleGameMaxPiece.link = undefined; // TODO: add game link support
+        this.singleGameMaxPiece = [gameLink];
```
Comment (EllAchE) [inline src/metrics/distances.ts:151]: why is this an array of gameLink? At the least the var name is confusing

File: src/metrics/misc.ts [hunk 10]
```diff
@@ -23,13 +23,14 @@ export function countGamesInDataset(datasetPath: string): number {
 
 export class MetadataMetric implements Metric {
```
Comment (EllAchE) [inline src/metrics/misc.ts:24]: maybe a quick top level comment on what this does

File: src/metrics/misc.ts [hunk 11]
```diff
@@ -207,39 +245,125 @@ export class MetadataMetric implements Metric {
...
+    // identify game endings
+    let gameEnd = metadata.find(item => item.startsWith('[Termination'));
+    const lastMove = game[game.length - 1].move;
+
+    if (gameEnd === '[Termination "Normal"]') {
+      if (lastMove.originalString.includes('#')) {
+        this.gameEndings['checkmate'] = (this.gameEndings['checkmate'] || 0) + 1;
+      } else if (result === '[Result "1/2-1/2"]') {
+        this.gameEndings['draw'] = (this.gameEndings['draw'] || 0) + 1;
+        if (this.chess.isStalemate()) {
```
Comment (EllAchE) [inline src/metrics/misc.ts:327]: I'm having trouble seeing where this.chess is instantiated with the board state. Recreating the board is also a little inefficient, but ik we need to understand the ending reason somehow, I'd need to look at the code again but maybe we can do it in a different place? I.e. when we get to the last move of a game history generator?

**Review by EllAchE — COMMENTED** [review]
I don't see any critical bugs. Wou;ld be nice to address/clean up some of my comments here though

**Comment by bennyrubanov** [comment]
merged this into #53 , closing it out

**Pull request closed without merging**